src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+        if (r < 0)
+                return r;
+
+        j->offline = true;
+
+        r = job_start(j);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to start job: %m");
+
+        j = NULL; /* Avoid job from being killed & freed */
+        return 1;
+}
+
+static BUS_DEFINE_PROPERTY_GET_ENUM(target_property_get_class, target_class, TargetClass);
+
+static int target_method_list_finish(
+                sd_bus_message *msg,
+                const Job *j,
+                JsonVariant *json,
+                sd_bus_error *error) {
+
+        JsonVariant *v;
+        _cleanup_strv_free_ char **versions = NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        int r;
+
+        assert(json);
+
+        /* Try our best since we have the data anyway, but isn't fatal if it fails */
+        (void) target_cache_version_from_json(j->target, json);
+
+        v = json_variant_by_key(json, "all");
+        if (!v)
+                return -EINVAL;
+
+        r = json_variant_strv(v, &versions);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_new_method_return(msg, &reply);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append_strv(reply, versions);
+        if (r < 0)
+                return r;
+
+        return sd_bus_send(NULL, reply, NULL);
+}
+
+static int target_method_list(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        Target *t = ASSERT_PTR(userdata);
+        _cleanup_(job_unrefp) Job *j = NULL;
+        bool offline;
+        int r;
+
+        assert(msg);
+
+        r = sd_bus_message_read(msg, "b", &offline);
+        if (r < 0)
+                return r;
+
+        const char *details[] = {
+                "class", target_class_to_string(t->class),
+                "name", t->name,
+                "offline", true_false(offline),
+                NULL
+        };
+
+        r = bus_verify_polkit_async(
+                msg,
+                CAP_SYS_ADMIN,
+                "org.freedesktop.sysupdate1.check",
+                details,
+                false,
+                UID_INVALID,
+                &j->manager->polkit_registry,
+                error);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return 1; /* Will call us back */
+
+        r = job_new(JOB_LIST, t, msg, target_method_list_finish, &j);
+        if (r < 0)
+                return r;
+
+        j->offline = offline;
+
+        r = job_start(j);
+        if (r < 0)
+                return sd_bus_error_set_errnof(error, r, "Failed to start job: %m");
+
+        j = NULL; /* Avoid job from being killed & freed */
+        return 1;
+}
+
+static int target_method_describe_finish(
+                sd_bus_message *msg,
+                const Job *j,
+                JsonVariant *json,
+                sd_bus_error *error) {
+        _cleanup_free_ char *reply;
```

Comment:
should be NULL initialized, so that if json_variant_format() fails you don't end up freeing a non-initialized pointer